lib/cluster/index.ts

```diff
@@ -424,6 +438,37 @@ class Cluster extends EventEmitter {
     return this.connectionPool.getNodes(role);
   }
 
+  // This is needed in order not to install a listener for each auto pipeline
+  public delayUntilReady(callback: CallbackFunction) {
+    // First call, setup the event listener
+    if (!this._readyDelayedCallbacks.length) {
+      this.once("ready", (...args) => {
+        for (const c of this._readyDelayedCallbacks) {
+          c(...args);
+        }
+
+        this._readyDelayedCallbacks = [];
+      });
```

Comment:
any chance that this may hang forever in case cluster becomes unresponsive for one reason or another? not saying that this should be happening in production, but lets say a command is queued, then cluster degrades, and is that state for multiple hours. this is, of course, a sign of a deeper issue, but the command will still be queued and potentially executed later. Should there be a mechanism to flush / remove the queued call in case of a fatal failure?

Reply:
That's a very good point, thanks for noticing.

First of all, let me clarify why I introduced this. With the new cluster auto pipeline routing, the `Redis.Cluster` instance needs to have slots-nodes information in order to assign commands to appropriate nodes group.
If these information are available (being up-to-date or not is not important thanks to `MOVED` and `ASK` support), the code above will never be invoked.

That fragment is only reached when using ioredis lazy connect feature (which is the default). Commands and/or pipelines can be prepared before the cluster server has sent slots information. So in that case I had to defer the execution until such info were ready. Thus, I listen for `ready` event.

I assume the `ready` even is only emitted once when the cluster is connected, isn't it? If that's the case, your concern above is correct, but it applies only when a Redis.Cluster instance is connected, not after. It's an edge case, but yet is possible.

So, in that case I should probably issue a `.once("error", ...)`, am I right?

Once you conferm it, I'll made the appropriate change.
Thanks!

Reply:
I'm not 100% sure about having an error listener here, likely similar logic to https://github.com/luin/ioredis/blob/ad5b45587b2e378c15fa879cc72580c391c3c18d/lib/redis/event_handler.ts#L154 can be used here. There is a separate close handler, which would abort transactions depending on the settings (having an offline queue, for instance) - so maybe create something similar - a prop, which would hold a reference to promise or something that could be rejected?

Reply:
It might work as well.

Another approach, by looking better at the code is to invoke callbacks with either success or failure internally in the chain in https://github.com/ShogunPanda/ioredis/blob/autopipelining/lib/cluster/index.ts#L195.

I think it's better as it is the only section in which those callbacks might be delayed.
What do you think?

Reply:
there was a bug with `connect()` when a cluster is not fully ready during the initial attempt - the promise would never resolve - more context here: https://github.com/luin/ioredis/pull/741

if you think that wont be a problem - then sure, an update can be added here

Reply:
Ok, I'll work on it and update the PR very soon.

Reply:
awesome :)

Reply:
Ok, I've updated it. Let me know if it works now.

Reply:
I think this looks awesome now 👍 @luin do you want to take a look and get this merged? 
I'm not sure whether this would warrant a breaking change release or not though

Reply:
@ShogunPanda @mcollina whats your personal opinion on changes - do they change the internals enough to potentially break existing integrations or I'm being overly cautious to be wanting to release this under 5.x.x?

Reply:
I'd go with semver minor. I do not expect regression - we are happy to jump on any that might result immediately.

Reply:
Confirming what Matteo said in terms of support, let me recap few things.

The autopipeline is completely opt-in and if not enabled the code is skipped and ioredis behaves at it was.
There are only two exceptions: 

1. Cluster pipelines are always delayed until slot informations are ready
2. Custom scripts are now always allowed in pipelines. In cluster mode, the scripts are loaded to the server before executing the pipeline.

Therefore, I also confirm this PR is compatible with semver minor release.